doc/developer/design/20211126-microbenchmarks.md

```diff
@@ -0,0 +1,135 @@
+
+For example, this microbenchmark will create a Kafka topic, ingest 100000 rows into it and measure the time it takes for the source to catch up.
+
+```
+class PartitionedKafkaIngestion(WallclockUntilSucces):
+   COUNT = 100000
+   def benchmark(c: Context):
+       return Benchmark(
+           setup = [
+               KafkaCreateTopic(partitions=5),
+               KafkaIngest('{"f1": 1} {"f2": 1}'),
+               CreateSource()
+           ],
+           benchmark = Sql("SELECT COUNT(*) = {COUNT} FROM {SOURCE}")
+       )
+```
+
+Internally, each of the helper classes, such as `KafkaCreateTopic` will produce a testdrive fragment such as `$ kafka-create-topic partitions=5` that will be passed on to testdrive for execution.
+
+### Running
+
+The execution will be via a `mzcompose`-based python script, something along the following lines:
+
+```
+cd test/microbenchmark
+./mzcompose run microbenchmark --compare-to materialized/latest
+```
+
+### Output
+
+The output will be something along those lines:
+
+```
+Microbenchmark            Iterations | Unit        | This:                 | Other:              | Outcome
+                                     |             | d45bbc0+modifications | materialized/latest |
+-----------------------------------------------------------------------------------------------------------------------------
+PartitionedKafkaIngestion         25 | records/sec | 11000                 | 10000               | This is 10% faster than Other
+...
+CountStar                            | ...         | ...                   | ...
+CountDistinct                        |             |
+OrderByLimitMaterialized             |             |
+OrderByLimitFinish                   |             |
+...
+```
+
+### Measurement strategies
+
+The measurement strategy describes exactly what will be measured and how.
+
+#### `Wallclock` strategy
+
+This strategy will measure the wallclock time between the last operation of the setup (the Kafka ingestion from the example), and the time the query under observation (the `SELECT COUNT(*)`) begins to return the expected result.
+
+For some queries, the result is available immediately without retrying, for example:
+
+```
+EXPLAIN SELECT * FROM t1 AS a1 LEFT JOIN t1 AS a2
+```
+
+or
+
+```
+SELECT COUNT(DISTINCT f1) FROM t1;
+```
+
+against a fully-ingested source.
+
+In our example above, the `SELECT` query will be retried until it returns the expected response. As `testdrive` is being used to execute the query, it will be instructed to 1) retry the query without an exponential back-off until it succeeds and 2) report wallclock execution times.
+
+#### `Counter` strategy
+
+This will be used in case the measurement needs to be obtained by querying a particular counter or parsing information in the `EXPLAIN`.
+
+### Benchmarking strategies
```

Comment:
From experience, the measurement method and strategy will need to be robust against the following:
- huge  variability in the results . Hopefully this will be handled by the fact that we will be looking at the minimum measured time, that is, maximum performance possible
- measurements having a bimodal distribution, where if a computation is completed within e.g. the kernel thread scheduling interval, it completes relatively fast, and it the thread is scheduled out, it completes slower. This causes the e.g. standard deviation to never converge to an acceptably low value, so can not be used to decide when the workload has been benchmarked enough times.
- if wallclock time is being used, overhead from steps unrelated to the operation being measured, for example, networking delays if copying large result sets over the TCP stack. For Materialize this may prove to be a substantial problem due to
A) lack of product-internal timers and counters that can be used in place of wallclock measurements
B) having to poll the materialized view to decide when it has "caught up", that is, processed all the data that was fed into the benchmark. We could use TAIL with PROGRESS and then measure "time until first TAIL progress marker" but testdrive is not currently set up to allow such a sequence of commands to pass through it
C) operations along the pipeline that are dependent on events that happen on a timer in some background thread. It would not be possible to obtain measurements with granularity better than those of the least granular timer involved.